connection_pool/round_robin.go

```diff
@@ -60,58 +61,40 @@ func (r *RoundRobinStrategy) IsEmpty() bool {
 	return r.size == 0
 }
 
-func (r *RoundRobinStrategy) CloseConns() []error {
-	r.mutex.Lock()
-	defer r.mutex.Unlock()
-
-	errs := make([]error, len(r.conns))
-
-	for i, conn := range r.conns {
-		errs[i] = conn.Close()
-	}
-
-	return errs
-}
-
 func (r *RoundRobinStrategy) GetNextConnection() *tarantool.Connection {
 	r.mutex.RLock()
 	defer r.mutex.RUnlock()
 
-	// We want to iterate through the elements in a circular order
-	// so the first element in cycle is connections[next]
-	// and the last one is connections[next + length].
-	next := r.nextIndex()
-	cycleLen := len(r.conns) + next
-	for i := next; i < cycleLen; i++ {
-		idx := i % len(r.conns)
-		if r.conns[idx].ConnectedNow() {
-			if i != next {
-				atomic.StoreUint64(&r.current, uint64(idx))
-			}
-			return r.conns[idx]
-		}
```

Comment:
Can you please highlight for me why this cycle is redundant?

Reply:
As far as I understand, the idea was to use only connected connections. But this code does not provide this guarantee: the connection can be closed immediately after the call:

1. thread1: conn := GetNextConnection()
2. thread2 (a goroutine in the connection): conn.closeConnection(err, true)
https://github.com/tarantool/go-tarantool/blob/master/connection.go#L672
3. thread1: conn.Select()

Although the probability of this event decreases, it cannot be avoided by this code.

Reply:
But does there any reason to increase the probability of receiving a dead connection?

Reply:
Yes, it increases the probability. But the situation itself cannot be avoided.

We have a logic to detect closed connections:

https://github.com/tarantool/go-tarantool/blob/oleg-jukovec/connection_pool_fixes/connection_pool/connection_pool.go#L712-L740

It's better to rely on it than to distribute useless checks around the code.

For example, it will further reduce the probability of this event. But that still doesn't guarantee anything:
```Go
func (pool *ConnectionPool) AnyFunc() {
     conn := pool.getNextConnection()
     if !conn.ClosedNow() {
         conn.Request()
     }
}
```

Reply:
Thank you for explanation!